conf/route_control.py

```diff
@@ -1,530 +1,851 @@
+from dataclasses import dataclass, field
+from threading import Lock, Thread
+from typing import Dict, List, Optional, Tuple
 
-# for retrieving neighbor info
+from pybess.bess import *
 from pyroute2 import IPDB, IPRoute
+from scapy.all import ICMP, IP, send
 
-from scapy.all import *
-
-try:
-    from pybess.bess import *
-except ImportError:
-    print('Cannot import the API module (pybess)')
-    raise
-
-MAX_RETRIES = 5
-SLEEP_S = 2
-
-
-class NeighborEntry:
-    def __init__(self):
-        self.neighbor_ip = None
-        self.iface = None
-        self.iprange = None
-        self.prefix_len = None
-        self.route_count = 0
-        self.gate_idx = 0
-        self.macstr = None
-
-    def __str__(self):
-        return ('{neigh: %s, iface: %s, ip-range: %s/%s}' %
-                (self.neighbor_ip, self.iface, self.iprange, self.prefix_len))
-
-
-def mac2hex(mac):
-    return int(mac.replace(':', ''), 16)
-
-
-def send_ping(neighbor_ip):
-    send(IP(dst=neighbor_ip) / ICMP())
-
+LOG_FORMAT = "%(asctime)s %(levelname)s %(message)s"
+logging.basicConfig(format=LOG_FORMAT, level=logging.INFO)
+logger = logging.getLogger(__name__)
 
-def send_arp(neighbor_ip, src_mac, iface):
-    pkt = Ether(dst="ff:ff:ff:ff:ff:ff") / ARP(pdst=neighbor_ip, hwsrc=src_mac)
-    pkt.show()
-    hexdump(pkt)
-    sendp(pkt, iface=iface)
+KEY_NETWORK_LAYER_DEST_ADDR = "NDA_DST"
+KEY_LINK_LAYER_ADDRESS = "NDA_LLADDR"
+KEY_NEW_NEIGHBOR_ACTION = "RTM_NEWNEIGH"
+KEY_DELETE_ROUTE_ACTION = "RTM_DELROUTE"
+KEY_NEW_ROUTE_ACTION = "RTM_NEWROUTE"
+KEY_INTERFACE = "RTA_OIF"
+KEY_DESTINATION_IP = "RTA_DST"
+KEY_DESTINATION_GATEWAY_IP = "RTA_GATEWAY"
+KEY_DESTINATION_PREFIX_LENGTH = "dst_len"
 
 
-def fetch_mac(dip):
-    ip = ''
-    _mac = ''
-    neighbors = ipr.get_neighbours(dst=dip)
-    for i in range(len(neighbors)):
-        for att in neighbors[i]['attrs']:
-            if 'NDA_DST' in att and dip == att[1]:
-                # ('NDA_DST', dip)
-                ip = att[1]
-            if 'NDA_LLADDR' in att:
-                # ('NDA_LLADDR', _mac)
-                _mac = att[1]
-                return _mac
+@dataclass
+class RouteEntry:
+    """A representation of a neighbor in route entry."""
+    next_hop_ip: str = field(default=None)
+    interface: str = field(default=None)
+    dest_prefix: str = field(default=None)
+    prefix_len: int = field(default=0)
 
 
-def link_modules(server, module, next_module, ogate=0, igate=0):
-    print('Linking {} module'.format(next_module))
-
-    # Pause bess first
-    bess.pause_all()
-    # Connect module to next_module
-    for _ in range(MAX_RETRIES):
-        try:
-            server.connect_modules(module, next_module, ogate, igate)
-        except BESS.Error as e:
-            bess.resume_all()
-            if e.code == errno.EBUSY:
-                break
+@dataclass
+class NeighborEntry:
+    """A representation of a neighbor in neighbor cache."""
+    gate_idx: int = field(default=0)
```

Comment:
Given that the "default" route (bad route) uses `MAX_GATES-1` as its default value. Should not `gate_idx` be set to `MAX_GATES-1`?

```suggestion
    gate_idx: int = field(default=MAX_GATES-1)
```

Reply:
currently it uses `get_gate_idx` to calculate it.
```
if (
            cached_entry := self._neighbor_cache.get(route_entry.next_hop_ip)
        ) is not None:
            return cached_entry.gate_idx
        return self._module_gate_count_cache[module_name]
```

Reply:
> Following the logic in the script before refactoring, `gate_idx` was never set to `MAX_GATES-1`. But if I understand you correctly Should `gate_idx` in the following code be `MAX_GATES-1` if the `prefix` is `0.0.0.0` and `prefix_len` is `0` ?
> 
> ```
> self._bess.run_module_command(
>                     module_name,
>                     "add",
>                     "IPLookupCommandAddArg",
>                     {
>                         "prefix": route_entry.dest_prefix,
>                         "prefix_len": route_entry.prefix_len,
>                         "gate": gate_idx,
>                     },
>                 )
> ```

Yes, please take a look at here: https://github.com/omec-project/upf/blob/master/conf/ports.py#L128